rfd/0229-agent-inventory-ui.md

```diff
@@ -0,0 +1,233 @@
+then begin populating itself with the agent instances and bot instances from the backend. Its reads from the backend will be rate-limited
+variably, based on the size of the cluster which will be derived from the size of the primary cache. Larger clusters (which can be assumed
+to have more bandwidth) will have more lenient rate-limits in order to prevent it from taking too long to populate.
+
+Aggregate counters such as the total number of up-to-date instances, total number of each service, etc. will be their own separate
+fields in the cache and tallied up during initialization, and updated as needed during create/edit/delete operations. All updater
+groups reported by agents will also be tallied into a deduplicated list.
+
+The cache will hook into the backend events watcher and watch for `instance` and `bot_instance` events and update the cache accordingly whenever
+an update is detected (such as an instance no longer being connected). This cache will live in auth and the data will be exposed to
+the proxy via a new RPC service with a `ListUnifiedInstances()` method which will build the page of instances by streaming from the
+`InventoryCache` sortcache.
+
+Any requests made for instances will be rejected until the `InventoryCache` is initialized and healthy, if not, all requests will be rejected
+and there will not be a fallback to the backend.
+
+#### UX
+
+![](assets/0229-agent-inventory-mockup.png)
+\*The above mockup is not finalized and may differ slightly from the implemented feature.
+
+The new `Instances Inventory` page will be available in the Web UI under the `Zero Trust Access` navigation section.
+
+The top of the page will contain aggregate metrics including the total number of up-to-date and out-of-date agents, number of
+agents using externally managed upgrades and their upgraders, and total active services.
+
+The list of instances will be a paginated list with infinite scroll support, and contain columns for hostname,
+version, services, and external upgrader (if any). A search bar will be available to use in either basic mode (default),
+which searches by hostname/intstance ID, or advanced mode using predicate language to perform queries. Where applicable, the predicate
+language query functions will be kept consistent with those that can be used on the Bot Instances dashboard.
+
+The `versions` UI filter control will contain 4 preset options:
+
+- `Up-to-date`: Filter for instances running the current version of the cluster (eg. `18.2.4`).
+- `Patch available`: Filter for instances running the same major version, but a lower minor version (eg. `>=18.0.0`)
+- `Upgrade required`: Filter for instances running 1 major version behind. (eg. `>=17.0.0 && <18.0.0`).
+- `Unsupported`: Filter for instances running more than 2 major versions behind, or more than 1 major version ahead. (eg. `<17.0.0 && >=19.0.0`)
```

Comment:
Can we include a preset(s) option for instances running less than or equal to some Teleport version or instances running a known vulnerable version of Teleport? For example, during the recent security incident customers were finding it difficult to identify any unpatched instances of Teleport. I realize this is easily achievable via a predicate expression from the advanced search option, but it seems to me that a preset for this case would be extremely valuable to a cluster admin.